library-best-practices/JLBP-4.md

```diff
@@ -0,0 +1,33 @@
+[JLBP-4] Avoid dependencies known to cause runtime conflicts
+------------------------------
+
+- Avoid using dependencies that are not stable.
+  - Example violation: The App Engine SDK introduced a non-final version of the
+    JCache API into its classpath, which conflicts with the final version of
+    JCache.
+  - The current exception is JSR 305 (`@NonNull`). It is widely referenced in
+    existing libraries and there is no compelling alternative, so it is allowed
+    as a dependency. However, it should also be noted that jsr305 has classes in
+    the javax.annotation package, which means that under the Java module system
+    introduced in Java 9, it conflicts with other artifacts that use that
+    package. The Error Prone team and others are in conversation with Jetbrains
+    and others to produce a good replacement.
+- Avoid using dependencies that duplicate classes with other dependencies.
+  - Most common case: a library that has published the same classes under
+    multiple artifact names. Always use the primary variant.
+  - Example: Guava's main artifact is `guava`, but from versions 13.0 to 17.0,
```

Comment:
I can't tell which of those ones conflict. Which should be the primary one to use?

Reply:
It depends on whether you want servlet 2.5, 3.0. 3.1, or 4.0. :-( 

javax.servlet:javax.servlet-api:3.1.0 for Java 8 on App Engine Standard
javax.servlet:servlet-api:2.5 for the Java & runtime

Reply:
This doesn't seem to be a concern for libraries though, right? (since our recommendation is to use the scope `provided`). Maybe the best place for this information could be the section on why `servlet-api` should be `provided`?